File: packages/sdk/src/administrator/manage-blueprints.ts
import BaseSDK from '../base-sdk';
import { QelosSDKOptions } from '../types';
import { IBlueprint } from '@qelos/global-types';

type IBlueprintUpdateParams = Omit<IBlueprint, 'tenant' | 'created' | 'updated'>;

export default class QlManageBlueprints extends BaseSDK {
  private relativePath = '/api/blueprints';

  constructor(options: QelosSDKOptions) {
    super(options)
  }

  getBlueprint(key: string) {
    return this.callJsonApi<IBlueprint>(`${this.relativePath}/${key}`)
  }

  getList() {
    return this.callJsonApi<IBlueprint[]>(this.relativePath);
  }

  remove(key: string): Promise<any> {
    return this.callApi(`${this.relativePath}/${key}`, { method: 'delete' });
  }

  update(key: string, changes: Partial<IBlueprintUpdateParams>): Promise<IBlueprint> {
    return this.callJsonApi<IBlueprint>(
      `${this.relativePath}/${key}`,
      {
        method: 'put',
        headers: { 'content-type': 'application/json' },
        body: JSON.stringify(changes)
      }
    )
  }

  create(blueprint: IBlueprintUpdateParams): Promise<IBlueprint> {
    return this.callJsonApi<IBlueprint>(this.relativePath, {
      method: 'post',
      headers: { 'content-type': 'application/json' },
      body: JSON.stringify(blueprint)
    })
  }
}

File: packages/python-sdk/qelos_sdk/administrator/manage_blueprints.py
from __future__ import annotations

import json
from typing import Any, Dict, List

from ..base_sdk import BaseSDK
from ..types import QelosSDKOptions


class QlManageBlueprints(BaseSDK):
    """Admin blueprint management: CRUD operations on blueprints."""

    _relative_path = "/api/blueprints"

    def __init__(self, options: QelosSDKOptions) -> None:
        super().__init__(options)

    async def get_blueprint(self, key: str) -> Dict[str, Any]:
        """Get a blueprint by key."""
        return await self.call_json_api(f"{self._relative_path}/{key}")

    async def get_list(self) -> List[Dict[str, Any]]:
        """List all blueprints."""
        return await self.call_json_api(self._relative_path)

    async def remove(self, key: str) -> Any:
        """Delete a blueprint by key."""
        return await self.call_api(f"{self._relative_path}/{key}", method="DELETE")

    async def update(self, key: str, changes: Dict[str, Any]) -> Dict[str, Any]:
        """Update a blueprint."""
        return await self.call_json_api(
            f"{self._relative_path}/{key}",
            method="PUT",
            headers={"content-type": "application/json"},
            body=json.dumps(changes),
        )

    async def create(self, blueprint: Dict[str, Any]) -> Dict[str, Any]:
        """Create a new blueprint."""
        return await self.call_json_api(
            self._relative_path,
            method="POST",
            headers={"content-type": "application/json"},
            body=json.dumps(blueprint),
        )
